lib/Infer/DataflowPruning.cpp

```diff
@@ -0,0 +1,335 @@
+      return Op0KB;
+    }
+    case souper::Inst::Xor : {
+      auto Op0KB = findKnownBits(I->Ops[0], C);
+      auto Op1KB = findKnownBits(I->Ops[1], C);
+      llvm::APInt KnownZeroOut =
+        (Op0KB.Zero & Op1KB.Zero) | (Op0KB.One & Op1KB.One);
+      Op0KB.One = (Op0KB.Zero & Op1KB.One) | (Op0KB.One & Op1KB.Zero);
+      Op0KB.Zero = std::move(KnownZeroOut);
+      // ^ logic copied from LLVM ValueTracking.cpp
+      return Op0KB;
+    }
+    default : return Result;
+  }
+}
+
+llvm::ConstantRange dataflow::findConstantRange(souper::Inst* I,
+                                                souper::ValueCache& C) {
+  llvm::ConstantRange result(I->Width);
+  switch (I->K) {
+    case souper::Inst::Const:
+    case souper::Inst::Var : {
+      EvalValue V = getValue(I, C);
+      if (V.hasValue()) {
+        return llvm::ConstantRange(V.getValue());
+      } else {
+        return result; // Whole range
+      }
+    }
+    case souper::Inst::Add: {
+      auto R0 = findConstantRange(I->Ops[0], C);
+      auto R1 = findConstantRange(I->Ops[1], C);
+      return R0.add(R1);
+    }
+    case souper::Inst::AddNSW: {
+      auto R0 = findConstantRange(I->Ops[0], C);
+      auto V1 = getValue(I->Ops[1], C);
+      if (V1.hasValue()) {
+        return R0.addWithNoSignedWrap(V1.getValue());
+      } else {
+        return result; // full range, can we do better?
+      }
+    }
+    case souper::Inst::Sub: {
+      auto R0 = findConstantRange(I->Ops[0], C);
+      auto R1 = findConstantRange(I->Ops[1], C);
+      return R0.sub(R1);
+    }
+    case souper::Inst::Mul: {
+      auto R0 = findConstantRange(I->Ops[0], C);
+      auto R1 = findConstantRange(I->Ops[1], C);
+      return R0.multiply(R1);
+    }
+    case souper::Inst::And: {
+      auto R0 = findConstantRange(I->Ops[0], C);
+      auto R1 = findConstantRange(I->Ops[1], C);
+      return R0.binaryAnd(R1);
+    }
+    case souper::Inst::Or: {
+      auto R0 = findConstantRange(I->Ops[0], C);
+      auto R1 = findConstantRange(I->Ops[1], C);
+      return R0.binaryOr(R1);
+    }
+    case souper::Inst::Shl: {
+      auto R0 = findConstantRange(I->Ops[0], C);
+      auto R1 = findConstantRange(I->Ops[1], C);
+      return R0.shl(R1);
+    }
+    case souper::Inst::AShr: {
+      auto R0 = findConstantRange(I->Ops[0], C);
+      auto R1 = findConstantRange(I->Ops[1], C);
+      return R0.ashr(R1);
+    }
+    case souper::Inst::LShr: {
+      auto R0 = findConstantRange(I->Ops[0], C);
+      auto R1 = findConstantRange(I->Ops[1], C);
+      return R0.lshr(R1);
+    }
+    case souper::Inst::UDiv: {
+      auto R0 = findConstantRange(I->Ops[0], C);
+      auto R1 = findConstantRange(I->Ops[1], C);
+      return R0.udiv(R1);
+    }
+//     case souper::Inst::SDiv: {
+//       auto R0 = FindConstantRange(I->Ops[0], C);
+//       auto R1 = FindConstantRange(I->Ops[1], C);
+//       return R0.sdiv(R1); // unimplemented
+//     }
+    // TODO: Xor pattern for not, truncs and extends, etc
```

Comment:
you don't have to do it for this PR but I think ext/trunc will increase the precision of your analysis quite a bit, since the transfer functions for these operations seem pretty precise